File: cli/xmltrncatorfixr.pl
#!/usr/bin/env perl
use strict;
use warnings;
use File::Copy;
use File::Basename;
use File::Spec;
use Getopt::Long;

my $VERSION = '1.0.0';

my $offset    = 50;
my $output    = '';
my $in_place  = 0;
my $quiet     = 0;
my $show_help = 0;
my $show_ver  = 0;

GetOptions(
    'offset|n=i' => \$offset,
    'output|o=s' => \$output,
    'in-place|i' => \$in_place,
    'quiet|q'    => \$quiet,
    'help|h'     => \$show_help,
    'version|V'  => \$show_ver,
) or usage(2);

if ($show_help) { usage(0); }
if ($show_ver)  { print "xmltrncatorfixr $VERSION\n"; exit 0; }

my $input = shift @ARGV;
usage(2) unless defined $input && length $input;
die "input file not found: $input\n" unless -f $input;

my $out_path = resolve_output($input, $output, $in_place);
copy($input, $out_path) or die "copy failed: $!\n";

my ($line, $col, $msg) = first_xml_error($out_path);
if (!defined $line) {
    info("no parser error detected; copy written to $out_path");
    exit 0;
}
info("first error at line $line, col $col: $msg");

my $byte = line_col_to_byte($out_path, $line, $col);
my $trunc_to = $byte - $offset;
$trunc_to = 0 if $trunc_to < 0;
truncate_file($out_path, $trunc_to);
info(sprintf("truncated to %d bytes (error byte %d minus offset %d)",
             $trunc_to, $byte, $offset));

my $rc = run_xmllint_recover($out_path);
if ($rc == 0) {
    info("xmllint --recover completed; output: $out_path");
} else {
    info("xmllint --recover returned $rc; partial output may still be usable");
}
exit 0;

sub resolve_output {
    my ($in, $out, $inplace) = @_;
    return $in if $inplace;
    return $out if length $out;
    my ($name, $dir, $ext) = fileparse($in, qr/\.[^.]*/);
    return File::Spec->catfile($dir || '.', "truncated-fixed-$name$ext");
}

sub which {
    my $exe = shift;
    my $sep = ($^O eq 'MSWin32') ? ';' : ':';
    my @ext = ($^O eq 'MSWin32') ? ('', '.exe', '.bat', '.cmd') : ('');
    for my $dir (split /\Q$sep\E/, ($ENV{PATH} || '')) {
        for my $e (@ext) {
            my $p = File::Spec->catfile($dir, $exe . $e);
            return $p if -x $p || ($^O eq 'MSWin32' && -e $p);
        }
    }
    return undef;
}

sub run_capture {
    my @cmd = @_;
    my $pid = open(my $fh, '-|');
    die "fork: $!" unless defined $pid;
    if ($pid == 0) {
        open(STDERR, '>&', \*STDOUT) or die;
        exec(@cmd) or exit 127;
    }
    local $/;
    my $out = <$fh>;
    close $fh;
    return ($? >> 8, defined $out ? $out : '');
}

sub first_xml_error {
    my $path = shift;
    my $xmllint = which('xmllint');
    die "xmllint not found in PATH (install libxml2-utils / libxml2)\n"
        unless defined $xmllint;
    my ($rc, $out) = run_capture($xmllint, '--noout', $path);
    return (undef, undef, undef) if $rc == 0;
    for my $line (split /\r?\n/, $out) {
        if ($line =~ /^\Q$path\E:(\d+):\s*(?:parser\s+error|error)\s*:\s*(.*)$/) {
            return ($1, 1, $2);
        }
        if ($line =~ /:(\d+):(\d+):\s*(.*)$/) {
            return ($1, $2, $3);
        }
    }
    return (undef, undef, $out);
}

sub line_col_to_byte {
    my ($path, $line, $col) = @_;
    open(my $fh, '<:raw', $path) or die "open $path: $!\n";
    my $byte = 0;
    my $cur_line = 1;
    while ($cur_line < $line) {
        my $c;
        my $r = read($fh, $c, 1);
        last unless $r;
        $byte++;
        $cur_line++ if $c eq "\n";
    }
    $byte += ($col > 0 ? $col - 1 : 0);
    close $fh;
    return $byte;
}

sub truncate_file {
    my ($path, $size) = @_;
    open(my $fh, '+<:raw', $path) or die "open $path: $!\n";
    truncate($fh, $size) or die "truncate: $!\n";
    close $fh;
}

sub run_xmllint_recover {
    my $path = shift;
    my $xmllint = which('xmllint');
    return 127 unless defined $xmllint;
    my $tmp = "$path.recover.tmp";
    my ($rc, undef) = run_capture($xmllint, '--recover', '--output', $tmp, $path);
    if (-s $tmp) {
        rename($tmp, $path) or die "rename: $!\n";
    } else {
        unlink $tmp;
    }
    return $rc;
}

sub info { print STDERR "[xmltrncatorfixr] @_\n" unless $quiet; }

sub usage {
    my $code = shift;
    my $fh = $code ? \*STDERR : \*STDOUT;
    print $fh <<'EOF';
Usage: xmltrncatorfixr [options] <input.xml>

Locates the first XML parser error, truncates the file just before the error
(minus a small safety offset), then runs `xmllint --recover` to repair the
closing tags. Useful for partial recovery of corrupt Office Open XML
(docx/xlsx/pptx) document parts.

Options:
  -n, --offset N    bytes to back off before the error (default: 50)
  -o, --output P    write to P (default: truncated-fixed-<input> beside input)
  -i, --in-place    overwrite the input file
  -q, --quiet       suppress progress messages
  -V, --version     print version
  -h, --help        show this help

Requires: xmllint (libxml2-utils on Debian/Ubuntu, libxml2 on macOS/Homebrew,
Windows builds available with libxml2 from xmlsoft.org).
EOF
    exit $code;
}

File: cli/xmltrncatorfixr.py
#!/usr/bin/env python3
"""xmltrncatorfixr - cross-platform XML truncator-fixer.

Locates the first XML parser error using expat (built into Python), truncates
the file just before that point (minus a safety offset), then runs `xmllint
--recover` to add proper closing tags. No third-party Python dependencies.
"""
from __future__ import annotations

import argparse
import os
import shutil
import subprocess
import sys
from pathlib import Path
from xml.parsers import expat

__version__ = "1.0.0"


def first_error_byte(path: Path) -> tuple[int | None, str]:
    parser = expat.ParserCreate()
    data = path.read_bytes()
    try:
        parser.Parse(data, True)
    except expat.ExpatError as exc:
        return parser.ErrorByteIndex, str(exc)
    return None, "no parser error"


def find_xmllint() -> str | None:
    return shutil.which("xmllint")


def run_xmllint_recover(path: Path) -> int:
    xmllint = find_xmllint()
    if not xmllint:
        return 127
    tmp = path.with_suffix(path.suffix + ".recover.tmp")
    rc = subprocess.run(
        [xmllint, "--recover", "--output", str(tmp), str(path)],
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
    ).returncode
    if tmp.exists() and tmp.stat().st_size > 0:
        tmp.replace(path)
    elif tmp.exists():
        tmp.unlink()
    return rc


def resolve_output(input_path: Path, output: str | None, in_place: bool) -> Path:
    if in_place:
        return input_path
    if output:
        return Path(output)
    return input_path.with_name(f"truncated-fixed-{input_path.name}")


def main(argv: list[str] | None = None) -> int:
    p = argparse.ArgumentParser(
        prog="xmltrncatorfixr",
        description="Truncate an XML file just before its first parser error "
        "and rebuild closing tags with xmllint --recover.",
    )
    p.add_argument("input", help="path to a (possibly broken) XML file")
    p.add_argument("-n", "--offset", type=int, default=50,
                   help="bytes to back off before the error (default: 50)")
    p.add_argument("-o", "--output", help="output path")
    p.add_argument("-i", "--in-place", action="store_true",
                   help="overwrite the input file")
    p.add_argument("-q", "--quiet", action="store_true",
                   help="suppress progress messages")
    p.add_argument("-V", "--version", action="version",
                   version=f"xmltrncatorfixr {__version__}")
    args = p.parse_args(argv)

    src = Path(args.input)
    if not src.is_file():
        print(f"input file not found: {src}", file=sys.stderr)
        return 2

    dst = resolve_output(src, args.output, args.in_place)
    if dst != src:
        shutil.copyfile(src, dst)

    err_byte, msg = first_error_byte(dst)
    log = (lambda *a: None) if args.quiet else (
        lambda *a: print("[xmltrncatorfixr]", *a, file=sys.stderr))

    if err_byte is None:
        log(f"no parser error detected; copy written to {dst}")
        return 0

    log(f"first error at byte {err_byte}: {msg}")
    trunc_to = max(0, err_byte - args.offset)
    with open(dst, "r+b") as fh:
        fh.truncate(trunc_to)
    log(f"truncated to {trunc_to} bytes (offset {args.offset})")

    rc = run_xmllint_recover(dst)
    if rc == 127:
        log("xmllint not found in PATH; truncated file written without recovery")
    elif rc == 0:
        log(f"xmllint --recover completed; output: {dst}")
    else:
        log(f"xmllint --recover returned {rc}; partial output may still be usable")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
